exonum-java-binding-common/src/test/java/com/exonum/binding/common/proofs/list/ListProofStructureValidatorTest.java

```diff
@@ -0,0 +1,228 @@
+        new ListProofBranch(leafOf(V2),
+            new ListProofHashNode(H3))
+    );
+
+    validator = createListProofStructureValidator(root);
+
+    assertThat(validator.getProofStatus(), is(ListProofStatus.INVALID_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_MaxAllowedDepth() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH;
+    ListProof root = generateRightLeaningProofTree(depth, leafOf(V1));
+
+    validator = createListProofStructureValidator(root);
+
+    assertTrue(validator.isValid());
+  }
+
+  @Test
+  void visit_UnbalancedElementNodeTooDeep() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH + 1;
+    ListProof root = generateRightLeaningProofTree(depth, leafOf(V1));
+
+    validator = createListProofStructureValidator(root);
+
+    assertThat(validator.getProofStatus(), is(ListProofStatus.INVALID_ELEMENT_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_UnbalancedHashNodeTooDeep() {
+    int depth = ListProofStructureValidator.MAX_NODE_DEPTH + 1;
+    ListProof root = generateRightLeaningProofTree(depth, new ListProofHashNode(H2));
+
+    validator = createListProofStructureValidator(root);
+
+    assertThat(validator.getProofStatus(), is(ListProofStatus.INVALID_HASH_NODE_DEPTH));
+  }
+
+  @Test
+  void visit_UnbalancedHashNodesOnlyLeafs() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            new ListProofHashNode(H1)
+        ),
+        new ListProofBranch(
+            new ListProofHashNode(H2), // <-- left leaf is hash node
+            new ListProofHashNode(H3)  // <-- right leaf is hash node
+        )
+    );
+
+    validator = createListProofStructureValidator(root);
+
+    assertThat(validator.getProofStatus(), is(ListProofStatus.INVALID_HASH_NODES_COUNT));
+  }
+
+  @Test
+  void visit_UnbalancedBranchHasOnlyHashNode() {
+    ListProofBranch root = new ListProofBranch(
+        new ListProofBranch(
+            leafOf(V1),
+            new ListProofHashNode(H1)
+        ),
+        new ListProofBranch(
+            new ListProofHashNode(H2), // <-- left leaf is hash node
+            null                 // <-- no right leaf
+        )
+    );
+
+    validator = createListProofStructureValidator(root);
+
+    assertThat(validator.getProofStatus(), is(ListProofStatus.INVALID_HASH_NODES_COUNT));
+  }
+
+  private ListProofStructureValidator createListProofStructureValidator(ListProof listProof) {
+    return new ListProofStructureValidator(listProof);
+  }
+
+  private static ListProofElement leafOf(String element) {
+    byte[] dbElement = bytesOf(element);
+    return new ListProofElement(dbElement);
+  }
+
+  private static byte[] bytesOf(String element) {
+    return StandardSerializers.string().toBytes(element);
+  }
+
+  private ListProof generateRightLeaningProofTree(int depth, ListProof leafNode) {
+    ListProof root = null;
+    ListProof left = leafNode;
+    int d = depth;
+    while (d != 0) {
```

Comment:
Shouldn't we assert beforehand that `d` is non-negative (or positive)?

Reply:
we can, but as for me it's redundant for test utils code.

Reply:
I'd agree — since it is `private` to this class.